### PR TITLE
clang-stdenv: move libstdcxxHook to cc-wrapper

### LINKED ISSUE #29204: llvmPackages_39.stdenv cannot find libstdc++
## Issue description

In Nixpkgs 17.09, `clang++` 3.9 cannot find libstdc++ headers:
```
$ nix-build -E 'with import <nixpkgs> {}; llvmPackages_39.stdenv.mkDerivation { name = "foo"; buildCommand = "echo \"#include <vector>\" > foo.cc; clang++ -c foo.cc -o $out"; }' -I nixpkgs=channel:nixos-17.09
these derivations will be built:
  /nix/store/m6cngdi7krryn25cm7bvgildpkjmi7bv-foo.drv
building path(s) '/nix/store/9c9dpbpdbkvpn1m8f6pgqzpygsxacw2j-foo'
building '/nix/store/m6cngdi7krryn25cm7bvgildpkjmi7bv-foo.drv'...
foo.cc:1:10: fatal error: 'vector' file not found
#include <vector>
         ^
1 error generated.
builder for '/nix/store/m6cngdi7krryn25cm7bvgildpkjmi7bv-foo.drv' failed with exit code 1
error: build of '/nix/store/m6cngdi7krryn25cm7bvgildpkjmi7bv-foo.drv' failed
```

This does work correctly with Clang 4.0 in 17.09, or Clang 3.9 in 17.03.

Setting `NIX_DEBUG=1` shows that the cc-wrapper is not adding the appropriate `-B` flags pointing to libstdc++ in gcc.

### Steps to reproduce

See above.

## Technical details

* System: NixOS 17.03
* Nix version: 1.12pre
* Nixpkgs version: 17.09beta285.b50193f0ca
* Sandboxing enabled: no

### PR DESCRIPTION
**Pull request by LnL7**
###### Motivation for this change

Fixes #29204

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] Linux
- [x] Tested via one or more nixpkgs test(s) (`tests.cc-wrapper`)
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/llvm/3.7/default.nix [hunk 1]
```diff
@@ -22,20 +25,35 @@ let
...
+    libstdcxxClang = ccWrapperFun {
+      cc = self.clang-unwrapped;
+      /* FIXME is this right? */
+      inherit (stdenv.cc) libc nativeTools nativeLibc;
+      extraPackages = [ libstdcxxHook ];
+    };
 
     libcxxClang = ccWrapperFun {
       cc = self.clang-unwrapped;
-      isClang = true;
```
Comment (Ericson2314) [inline pkgs/development/compilers/llvm/3.7/default.nix]: Is this because it figures it out automatically?

**Review by Ericson2314 — APPROVED** [review]
LGTM!

File: pkgs/development/compilers/llvm/3.7/default.nix (hunk 1 above)
Comment (LnL7) [reply pkgs/development/compilers/llvm/3.7/default.nix]: Yeah, it's not necessary.

**Comment by corngood** [comment]
I bisected a problem I'm having in one of my local environments to this commit.

```
$ nix-shell -p gcc -p clang --command 'echo "int main() { return 0; }" | c++ -x c++ -'
g++: error: unrecognized command line option ‘-stdlib=libstdc++’; did you mean ‘-static-libstdc++’?
```

I only just started looking into it, but I thought I'd check if anyone had any thoughts.  I'm creating an environment with gcc for building, and clang for tooling.

**Comment by LnL7** [comment]
@corngood See https://github.com/NixOS/nixpkgs/issues/29877 for discussion on that problem.

**Pull request merged**